syntax = "proto3";

package boltz;

import "common.proto";

// Training service
service TrainingService {
  // Start training job
  rpc StartTraining (TrainingRequest) returns (TrainingResponse) {}
  
  // Get training job status
  rpc GetJobStatus (JobStatusRequest) returns (TrainingJobStatusResponse) {}
  
  // Cancel training job
  rpc CancelJob (CancelJobRequest) returns (CancelJobResponse) {}

  // Export model
  rpc ExportModel (ExportModelRequest) returns (ExportModelResponse) {}
}

// Request for training
message TrainingRequest {
  string job_id = 1;
  string config_path = 2;
  repeated string args = 3;
  int32 num_gpus = 4;
  string output_dir = 5;
  bool resume = 6;
  string checkpoint = 7;
  string experiment_name = 8;
  map<string, string> hyperparameters = 9;
}

// Response for training
message TrainingResponse {
  string job_id = 1;
  string status = 2;
  string checkpoint_path = 3;
  string error_message = 4;
}

// Extended job status response for training
message TrainingJobStatusResponse {
  JobStatusResponse base = 1;
  float current_epoch = 2;
  float val_loss = 3;
  float train_loss = 4;
  string checkpoint_path = 5;
}

// Request to export model
message ExportModelRequest {
  string job_id = 1;
  string output_path = 2;
  string format = 3;
}

// Response for model export
message ExportModelResponse {
  string job_id = 1;
  string status = 2;
  string model_path = 3;
  string error_message = 4;
}
